README.md
# 🏙 Tokyo Night

A dark and light Neovim theme written in Lua ported from the Visual Studio Code
[TokyoNight](https://github.com/enkia/tokyo-night-vscode-theme) theme. Includes
extra themes for Kitty, Alacritty, iTerm and Fish.

## Storm

![image](https://user-images.githubusercontent.com/292349/115295095-3a9e5080-a10e-11eb-9aed-6054488c46ce.png)

## Night

![image](https://user-images.githubusercontent.com/292349/115295327-7afdce80-a10e-11eb-89b3-2591262bf95a.png)

## Moon

![image](https://user-images.githubusercontent.com/292349/190951628-10ba28a1-57ff-4479-8eab-47400a402242.png)

## Day

![image](https://user-images.githubusercontent.com/292349/115996270-78c6c480-a593-11eb-8ed0-7d1400b058f5.png)

## ✨ Features

- supports the latest Neovim 0.9.0 features
- terminal colors
- darker background for sidebar-like windows
- supports all major plugins
- TokyoNight [extras](#-extras) for a lot of other apps

### 🍭 Extras

<!-- extras:start -->
- [Alacritty](https://github.com/alacritty/alacritty) ([alacritty](extras/alacritty))
- [Delta](https://github.com/dandavison/delta) ([delta](extras/delta))
- [Fish](https://fishshell.com/docs/current/index.html) ([fish](extras/fish))
- [Fish Themes](https://fishshell.com/docs/current/interactive.html#syntax-highlighting) ([fish_themes](extras/fish_themes))
- [Foot](https://codeberg.org/dnkl/foot) ([foot](extras/foot))
- [iTerm](https://iterm2.com/) ([iterm](extras/iterm))
- [Kitty](https://sw.kovidgoyal.net/kitty/conf.html) ([kitty](extras/kitty))
- [Lua Table for testing](https://www.lua.org) ([lua](extras/lua))
- [Prism](https://prismjs.com) ([prism](extras/prism))
- [Sublime Text](https://www.sublimetext.com/docs/themes) ([sublime](extras/sublime))
- [Terminator](https://gnome-terminator.readthedocs.io/en/latest/config.html) ([terminator](extras/terminator))
- [Tilix](https://github.com/gnunn1/tilix) ([tilix](extras/tilix))
- [Tmux](https://github.com/tmux/tmux/wiki) ([tmux](extras/tmux))
- [WezTerm](https://wezfurlong.org/wezterm/config/) ([wezterm](extras/wezterm))
- [Windows Terminal](https://aka.ms/terminal-documentation) ([windows_terminal](extras/windows_terminal))
- [Xfce Terminal](https://docs.xfce.org/apps/terminal/advanced) ([xfceterm](extras/xfceterm))
- [Xresources](https://wiki.archlinux.org/title/X_resources) ([xresources](extras/xresources))
<!-- extras:end -->

## ⚡️ Requirements

- Neovim >= 0.6.0

## 📦 Installation

Install the theme with your preferred package manager:

[vim-plug](https://github.com/junegunn/vim-plug)

```vim
Plug 'folke/chibanight.nvim', { 'branch': 'main' }
```

[packer](https://github.com/wbthomason/packer.nvim)

```lua
use 'folke/chibanight.nvim'
```

## 🚀 Usage

Enable the colorscheme:

```vim
" Vim Script
colorscheme chibanight

" There are also colorschemes for the different styles
colorscheme chibanight-night
colorscheme chibanight-storm
colorscheme chibanight-day
colorscheme chibanight-moon
```

```lua
-- Lua
vim.cmd[[colorscheme chibanight]]
```

To enable the `chibanight` theme for `Barbecue`:

```lua
require('barbecue').setup {
  -- ... your barbecue config
  theme = 'chibanight',
  -- ... your barbecue config
}
```

To enable the `TokyoNight` theme for `Lualine`, simply specify it in your
lualine settings:

```lua
require('lualine').setup {
  options = {
    -- ... your lualine config
    theme = 'chibanight'
    -- ... your lualine config
  }
}
```

To enable the `chibanight` colorscheme for `Lightline`:

```vim
" Vim Script
let g:lightline = {'colorscheme': 'chibanight'}
```

## ⚙️ Configuration

> ❗️ configuration needs to be set **BEFORE** loading the color scheme with
> `colorscheme chibanight`

The theme comes in four styles, `storm`, `moon`, a darker variant `night` and `day`.

The **day** style will be used if:

- `{ style = "day"}` was passed to `setup(options)`
- or `vim.o.background = "light"`

TokyoNight will use the default options, unless you call `setup`.

```lua
require("chibanight").setup({
  -- your configuration comes here
  -- or leave it empty to use the default settings
  style = "storm", -- The theme comes in three styles, `storm`, `moon`, a darker variant `night` and `day`
  light_style = "day", -- The theme is used when the background is set to light
  transparent = false, -- Enable this to disable setting the background color
  terminal_colors = true, -- Configure the colors used when opening a `:terminal` in Neovim
  styles = {
    -- Style to be applied to different syntax groups
    -- Value is any valid attr-list value for `:help nvim_set_hl`
    comments = { italic = true },
    keywords = { italic = true },
    functions = {},
    variables = {},
    -- Background styles. Can be "dark", "transparent" or "normal"
    sidebars = "dark", -- style for sidebars, see below
    floats = "dark", -- style for floating windows
  },
  sidebars = { "qf", "help" }, -- Set a darker background on sidebar-like windows. For example: `["qf", "vista_kind", "terminal", "packer"]`
  day_brightness = 0.3, -- Adjusts the brightness of the colors of the **Day** style. Number between 0 and 1, from dull to vibrant colors
  hide_inactive_statusline = false, -- Enabling this option, will hide inactive statuslines and replace them with a thin border instead. Should work with the standard **StatusLine** and **LuaLine**.
  dim_inactive = false, -- dims inactive windows
  lualine_bold = false, -- When `true`, section headers in the lualine theme will be bold

  --- You can override specific color groups to use other groups or a hex color
  --- function will be called with a ColorScheme table
  ---@param colors ColorScheme
  on_colors = function(colors) end,

  --- You can override specific highlights to use other groups or a hex color
  --- function will be called with a Highlights and ColorScheme table
  ---@param highlights Highlights
  ---@param colors ColorScheme
  on_highlights = function(highlights, colors) end,
})
```

## 🪓 Overriding Colors & Highlight Groups

How the highlight groups are calculated:

1. the **colors** for the style are calculated based on your config
2. `config.on_colors(colors)` is ran, where you can override the colors
3. the **colors** are then used to generate the highlight groups
4. `config.on_highlights(highlights, colors)` is ran, where you can overide the highlight groups

Please refer to default values for `colors` and `highlights` for the [storm](extras/lua/chibanight_storm.lua), [moon](extras/lua/chibanight_moon.lua), [night](extras/lua/chibanight_night.lua), [day](extras/lua/chibanight_day.lua)

Example for changing some settings and colors

```lua
require("chibanight").setup({
  -- use the night style
  style = "night",
  -- disable italic for functions
  styles = {
    functions = {}
  },
  sidebars = { "qf", "vista_kind", "terminal", "packer" },
  -- Change the "hint" color to the "orange" color, and make the "error" color bright red
  on_colors = function(colors)
    colors.hint = colors.orange
    colors.error = "#ff0000"
  end
})
```

Example to make Telescope
[borderless](https://github.com/nvim-telescope/telescope.nvim/wiki/Gallery#borderless)

```lua
require("chibanight").setup({
  on_highlights = function(hl, c)
    local prompt = "#2d3149"
    hl.TelescopeNormal = {
      bg = c.bg_dark,
      fg = c.fg_dark,
    }
    hl.TelescopeBorder = {
      bg = c.bg_dark,
      fg = c.bg_dark,
    }
    hl.TelescopePromptNormal = {
      bg = prompt,
    }
    hl.TelescopePromptBorder = {
      bg = prompt,
      fg = prompt,
    }
    hl.TelescopePromptTitle = {
      bg = prompt,
      fg = prompt,
    }
    hl.TelescopePreviewTitle = {
      bg = c.bg_dark,
      fg = c.bg_dark,
    }
    hl.TelescopeResultsTitle = {
      bg = c.bg_dark,
      fg = c.bg_dark,
    }
  end,
})
```

### Making `undercurls` work properly in **Tmux**

To have undercurls show up and in color, add the following to your **Tmux**
config file:

```sh
# Undercurl
set -g default-terminal "${TERM}"
set -as terminal-overrides ',*:Smulx=\E[4::%p1%dm'  # undercurl support
set -as terminal-overrides ',*:Setulc=\E[58::2::%p1%{65536}%/%d::%p1%{256}%/%{255}%&%d::%p1%{255}%&%d%;m'  # underscore colours - needs tmux-3.0
```

## 🍭 Extras

Extra color configs for **Kitty**, **Alacritty**, **Fish**, **WezTerm**,
**iTerm** and **foot** can be found in [extras](extras/). To use them, refer to
their respective documentation.

![image](https://user-images.githubusercontent.com/292349/115395546-d8d6f880-a198-11eb-98fb-a1194787701d.png)

You can easily use the color palette for other plugins inside your Neovim
config:

```lua
local colors = require("chibanight.colors").setup() -- pass in any of the config options as explained above
local util = require("chibanight.util")

aplugin.background = colors.bg_dark
aplugin.my_error = util.lighten(colors.red1, 0.3) -- number between 0 and 1. 0 results in white, 1 results in red1
```

## 🔥 Contributing

Pull requests are welcome. For the `extras`, we use a simple template system
that can be used to generate themes for the different styles.

How to add a new extra template:

1. create a file like `lua/chibanight/extra/cool-app.lua`
2. add the name and output file extension to the `extras` table in
   `lua/chibanight/extra/init.lua`
3. run the shell script below to generate / update
   extra themes

   ```sh
   $ nvim --headless "+lua require('chibanight.extra').setup()" +qa
   ```

4. check the newly created themes under `extra/`, but **DO NOT** commit them! They will be build automatically by the ci


CHANGELOG.md
# Changelog

## [1.18.0](https://github.com/folke/chibanight.nvim/compare/v1.17.0...v1.18.0) (2023-05-07)


### Features

* **lsp:** add more highlights ([#374](https://github.com/folke/chibanight.nvim/issues/374)) ([fc4b07d](https://github.com/folke/chibanight.nvim/commit/fc4b07d1585396ee14f6d6323fd534dfd5526ebe))
* **tsx:** better colors for tsx tags ([f10213d](https://github.com/folke/chibanight.nvim/commit/f10213d25ed5341546a9d99ad4ff4a4612200cb5))


### Bug Fixes

* **GitGutter:** add highlight of GitGutter**LineNr ([#367](https://github.com/folke/chibanight.nvim/issues/367)) ([606d4d9](https://github.com/folke/chibanight.nvim/commit/606d4d9e7c3c95ae9243d652ce6c228958fb395b))

## [1.17.0](https://github.com/folke/chibanight.nvim/compare/v1.16.0...v1.17.0) (2023-04-22)


### Features

* added extras to readme ([cca8227](https://github.com/folke/chibanight.nvim/commit/cca8227dc4c6e0c4f5a56055b659f3c04efe496f))


### Bug Fixes

* **colors:** make extras deterministic. Fixes [#344](https://github.com/folke/chibanight.nvim/issues/344) ([3f18207](https://github.com/folke/chibanight.nvim/commit/3f18207ff1f1c6e68d292f4e6c804dc8eefd6b2f))
* **delta:** move color defs to colors so it works properly with the day theme. Fixes [#362](https://github.com/folke/chibanight.nvim/issues/362) ([d78d3de](https://github.com/folke/chibanight.nvim/commit/d78d3de58c15c8369e920b0852be3dc81797e331))

## [1.16.0](https://github.com/folke/chibanight.nvim/compare/v1.15.0...v1.16.0) (2023-04-19)


### Features

* Improve night barbar ([#360](https://github.com/folke/chibanight.nvim/issues/360)) ([4a5ea8c](https://github.com/folke/chibanight.nvim/commit/4a5ea8c5df533ca92b590c31d4a3930562a3a257))
* **semantic_tokens:** highlight lsp.typemod.macro.defaultLibrary as `[@function](https://github.com/function).builtin` ([#358](https://github.com/folke/chibanight.nvim/issues/358)) ([dc5b5e2](https://github.com/folke/chibanight.nvim/commit/dc5b5e20bc232a10e833fd3e81fea149bd4a902f))

## [1.15.0](https://github.com/folke/chibanight.nvim/compare/v1.14.0...v1.15.0) (2023-04-19)


### Features

* **treesitter:** darken builtin types & revert reset comment type ([#356](https://github.com/folke/chibanight.nvim/issues/356)) ([b6f207e](https://github.com/folke/chibanight.nvim/commit/b6f207ebf6e3de9acbbc443623ad74b65bd00d05))

## [1.14.0](https://github.com/folke/chibanight.nvim/compare/v1.13.0...v1.14.0) (2023-04-18)


### Features

* better Debug colors ([a1c8489](https://github.com/folke/chibanight.nvim/commit/a1c8489cb6116d3c237be71d0a7dd96fbc982d91))


### Bug Fixes

* **semantic_tokens:** reset comment type ([#354](https://github.com/folke/chibanight.nvim/issues/354)) ([3d2f79c](https://github.com/folke/chibanight.nvim/commit/3d2f79cd1f3cea7520d5cc4d06c8010fe2592db8))

## [1.13.0](https://github.com/folke/chibanight.nvim/compare/v1.12.0...v1.13.0) (2023-04-18)


### Features

* better style for DiagnosticUnnecessary ([6ebd358](https://github.com/folke/chibanight.nvim/commit/6ebd35868fbdb8f5a3c5009d5d7a72ed7da2143c))
* **semantic_tokens:** slightly different style for interfaces and default types ([3af1264](https://github.com/folke/chibanight.nvim/commit/3af126456719cdc6c2a904cffc3cfe31ce22afb6))


### Bug Fixes

* **semantic_tokens:** slightly less difference for builtin types ([939bd8c](https://github.com/folke/chibanight.nvim/commit/939bd8c7e149391b0cb71d3aadd75eba0fa5647f))

## [1.12.0](https://github.com/folke/chibanight.nvim/compare/v1.11.0...v1.12.0) (2023-04-17)


### Features

* improve barbar separator ([#350](https://github.com/folke/chibanight.nvim/issues/350)) ([169b76f](https://github.com/folke/chibanight.nvim/commit/169b76ff87cab82d8691414db9807cd1ae9e560e))


### Bug Fixes

* **lsp:** enumMember should be highlighted `[@constant](https://github.com/constant)` ([#352](https://github.com/folke/chibanight.nvim/issues/352)) ([78bcc2f](https://github.com/folke/chibanight.nvim/commit/78bcc2fae719134291df218f679ec2bb49c6e95f))

## [1.11.0](https://github.com/folke/chibanight.nvim/compare/v1.10.0...v1.11.0) (2023-04-16)


### Features

* added support for LazyVim DapStoppedLine ([ef6df2f](https://github.com/folke/chibanight.nvim/commit/ef6df2f2d34d1cdc8c5653db3d1b8d454d844858))
* better hl for markdown literal ([e061044](https://github.com/folke/chibanight.nvim/commit/e0610445d5d2ee089269a199f49bbfaebd1c9ced))


### Bug Fixes

* **barbar:** CurrentBackground ([#342](https://github.com/folke/chibanight.nvim/issues/342)) ([cbed164](https://github.com/folke/chibanight.nvim/commit/cbed164e6f280eced47de91107a87ede5fc2adcf))
* **extras:** make generation of extras deterministic ([#344](https://github.com/folke/chibanight.nvim/issues/344)) ([715f923](https://github.com/folke/chibanight.nvim/commit/715f923435a9255f85dec8380fac027f37acac72))
* updated extras ([6e0bfd6](https://github.com/folke/chibanight.nvim/commit/6e0bfd6836e09a4c66184a33f1266e964f920299))

## [1.10.0](https://github.com/folke/chibanight.nvim/compare/v1.9.1...v1.10.0) (2023-03-23)


### Features

* **lsp:** set injected highlights ([#340](https://github.com/folke/chibanight.nvim/issues/340)) ([3b33146](https://github.com/folke/chibanight.nvim/commit/3b331463ed4f88ff8052a054f3e898bcc1a638bd))

## [1.9.1](https://github.com/folke/chibanight.nvim/compare/v1.9.0...v1.9.1) (2023-03-19)


### Bug Fixes

* **alpha:** AlphaFooter ([8be0a63](https://github.com/folke/chibanight.nvim/commit/8be0a63987d43a4ae0fc2420fd5100fa528a92ff))

## [1.9.0](https://github.com/folke/chibanight.nvim/compare/v1.8.0...v1.9.0) (2023-03-14)


### Features

* **semantic_tokens:** link lsp comments, e.g. conditional C ifdefs that won't evaluate ([#334](https://github.com/folke/chibanight.nvim/issues/334)) ([b233a3c](https://github.com/folke/chibanight.nvim/commit/b233a3cab08699c73239346cd7d51a2a55c62250))

## [1.8.0](https://github.com/folke/chibanight.nvim/compare/v1.7.0...v1.8.0) (2023-03-12)


### Features


lua/chibanight/util.lua
local ts = require("chibanight.treesitter")

local M = {}

M.bg = "#000000"
M.fg = "#ffffff"
M.day_brightness = 0.3

---@param c  string
local function hexToRgb(c)
  c = string.lower(c)
  return { tonumber(c:sub(2, 3), 16), tonumber(c:sub(4, 5), 16), tonumber(c:sub(6, 7), 16) }
end

---@param foreground string foreground color
---@param background string background color
---@param alpha number|string number between 0 and 1. 0 results in bg, 1 results in fg
function M.blend(foreground, background, alpha)
  alpha = type(alpha) == "string" and (tonumber(alpha, 16) / 0xff) or alpha
  local bg = hexToRgb(background)
  local fg = hexToRgb(foreground)

  local blendChannel = function(i)
    local ret = (alpha * fg[i] + ((1 - alpha) * bg[i]))
    return math.floor(math.min(math.max(0, ret), 255) + 0.5)
  end

  return string.format("#%02x%02x%02x", blendChannel(1), blendChannel(2), blendChannel(3))
end

function M.darken(hex, amount, bg)
  return M.blend(hex, bg or M.bg, amount)
end

function M.lighten(hex, amount, fg)
  return M.blend(hex, fg or M.fg, amount)
end

function M.invert_color(color)
  local hsluv = require("chibanight.hsluv")
  if color ~= "NONE" then
    local hsl = hsluv.hex_to_hsluv(color)
    hsl[3] = 100 - hsl[3]
    if hsl[3] < 40 then
      hsl[3] = hsl[3] + (100 - hsl[3]) * M.day_brightness
    end
    return hsluv.hsluv_to_hex(hsl)
  end
  return color
end

---@param group string
function M.highlight(group, hl)
  group = ts.get(group)
  if not group then
    return
  end
  if hl.style then
    if type(hl.style) == "table" then
      hl = vim.tbl_extend("force", hl, hl.style)
    elseif hl.style:lower() ~= "none" then
      -- handle old string style definitions
      for s in string.gmatch(hl.style, "([^,]+)") do
        hl[s] = true
      end
    end
    hl.style = nil
  end
  vim.api.nvim_set_hl(0, group, hl)
end

--- Delete the autocmds when the theme changes to something else
function M.onColorScheme()
  vim.cmd([[autocmd! TokyoNight]])
  vim.cmd([[augroup! TokyoNight]])
end

---@param config Config
function M.autocmds(config)
  vim.cmd([[augroup TokyoNight]])
  vim.cmd([[  autocmd!]])
  vim.cmd([[  autocmd ColorSchemePre * lua require("chibanight.util").onColorScheme()]])

  vim.cmd(
    [[  autocmd FileType ]]
      .. table.concat(config.sidebars, ",")
      .. [[ setlocal winhighlight=Normal:NormalSB,SignColumn:SignColumnSB]]
  )
  if vim.tbl_contains(config.sidebars, "terminal") then
    vim.cmd([[  autocmd TermOpen * setlocal winhighlight=Normal:NormalSB,SignColumn:SignColumnSB]])
  end
  vim.cmd([[augroup end]])
end

-- Simple string interpolation.
--
-- Example template: "${name} is ${value}"
--
---@param str string template string
---@param table table key value pairs to replace in the string
function M.template(str, table)
  return (
    str:gsub("($%b{})", function(w)
      return vim.tbl_get(table, unpack(vim.split(w:sub(3, -2), ".", { plain = true }))) or w
    end)
  )
end

function M.syntax(syntax)
  for group, colors in pairs(syntax) do
    M.highlight(group, colors)
  end
end

---@param colors ColorScheme
function M.terminal(colors)
  -- dark
  vim.g.terminal_color_0 = colors.black
  vim.g.terminal_color_8 = colors.terminal_black

  -- light
  vim.g.terminal_color_7 = colors.fg_dark
  vim.g.terminal_color_15 = colors.fg

  -- colors
  vim.g.terminal_color_1 = colors.red
  vim.g.terminal_color_9 = colors.red

  vim.g.terminal_color_2 = colors.green
  vim.g.terminal_color_10 = colors.green

  vim.g.terminal_color_3 = colors.yellow
  vim.g.terminal_color_11 = colors.yellow

  vim.g.terminal_color_4 = colors.blue
  vim.g.terminal_color_12 = colors.blue

  vim.g.terminal_color_5 = colors.magenta
  vim.g.terminal_color_13 = colors.magenta

  vim.g.terminal_color_6 = colors.cyan
  vim.g.terminal_color_14 = colors.cyan
end

---@param colors ColorScheme
function M.invert_colors(colors)
  if type(colors) == "string" then
    ---@diagnostic disable-next-line: return-type-mismatch
    return M.invert_color(colors)
  end
  for key, value in pairs(colors) do
    colors[key] = M.invert_colors(value)
  end
  return colors
end

---@param hls Highlights
function M.invert_highlights(hls)
  for _, hl in pairs(hls) do
    if hl.fg then
      hl.fg = M.invert_color(hl.fg)
    end
    if hl.bg then
      hl.bg = M.invert_color(hl.bg)
    end
    if hl.sp then
      hl.sp = M.invert_color(hl.sp)
    end
  end
end

---@param theme Theme
function M.load(theme)
  -- only needed to clear when not the default colorscheme
  if vim.g.colors_name then
    vim.cmd("hi clear")
  end

  vim.o.termguicolors = true
  vim.g.colors_name = "chibanight"

  if ts.new_style() then
    for group, colors in pairs(ts.defaults) do
      if not theme.highlights[group] then
        M.highlight(group, colors)
      end
    end
  end

  M.syntax(theme.highlights)

  -- vim.api.nvim_set_hl_ns(M.ns)
  if theme.config.terminal_colors then
    M.terminal(theme.colors)
  end

  M.autocmds(theme.config)

  vim.defer_fn(function()
    M.syntax(theme.defer)
  end, 100)
end

return M
